server/src/main.rs

```diff
@@ -125,174 +103,174 @@ impl LanguageServer for Backend {
     }
 
     async fn hover(&self, params: HoverParams) -> tower_lsp::jsonrpc::Result<Option<Hover>> {
-        let uri = Url::from_file_path(
-            params
-                .text_document_position_params
-                .text_document
-                .uri
-                .to_file_path()
-                .unwrap(),
-        )
-        .unwrap();
+        let uri = params.text_document_position_params.text_document.uri;
         let position = params.text_document_position_params.position;
-        let compilation_state = &self.compilation_state.lock().await;
-        Ok(
-            get_hover_info(compilation_state, uri, position).map(|info| Hover {
-                contents: HoverContents::Scalar(MarkedString::String(info)),
-                range: None,
-            }),
-        )
+
+        // Convert the URI to a file path and back to a URL to ensure that the URI is formatted correctly for Windows.
+        let file_name = url_to_file_path(&uri).ok_or_else(Error::internal_error)?;
+        let url = uri
+            .to_file_path()
+            .and_then(Url::from_file_path)
+            .map_err(|_| Error::internal_error())?;
+
+        // Find the configuration set that contains the file and get the hover info
+        let configuration_sets = self.session.configuration_sets.lock().await;
+        configuration_sets
+            .iter()
+            .find_file(&file_name)
+            .map(|compilation_state| try_into_hover_result(compilation_state, url, position))
+            .transpose()
     }
 
-    async fn did_open(&self, _: DidOpenTextDocumentParams) {
-        self.handle_file_change().await;
+    async fn did_open(&self, params: DidOpenTextDocumentParams) {
+        if let Some(file_name) = url_to_file_path(&params.text_document.uri) {
+            self.handle_file_change(&file_name).await;
+        }
     }
 
-    async fn did_save(&self, _: DidSaveTextDocumentParams) {
-        self.client
-            .log_message(MessageType::INFO, "file saved")
-            .await;
-        self.handle_file_change().await;
+    async fn did_save(&self, params: DidSaveTextDocumentParams) {
+        if let Some(file_name) = url_to_file_path(&params.text_document.uri) {
+            self.handle_file_change(&file_name).await;
+        }
     }
 }
 
 impl Backend {
-    async fn handle_file_change(&self) {
-        let updated_state = self.compile_slice_files().await;
-
-        let mut compilation_state_lock = self.compilation_state.lock().await;
-        *compilation_state_lock = updated_state;
-
-        self.publish_diagnostics_for_all_files(&mut compilation_state_lock)
-            .await;
-    }
-
-    async fn compile_slice_files(&self) -> CompilationState {
+    async fn handle_file_change(&self, file_name: &str) {
         self.client
-            .log_message(MessageType::INFO, "compiling slice")
+            .log_message(MessageType::INFO, format!("File {} changed", file_name))
```

Comment:
Quotes would make it a little easier to read where the name starts/stops, also, nicer format string:
```suggestion
            .log_message(MessageType::INFO, format!("File '{file_name}' changed"))
```